internal/discovery/dkv_discovery_service.go

```diff
@@ -0,0 +1,130 @@
+	return &DiscoveryConfig{statusTTL, heartbeatTimeout}
+}
+
+type discoverService struct {
+	// TODO - dkvServer should point to interface of master (currently distributedService)
+	dkvServer DKVServer
+	logger    *zap.Logger
+	config    *DiscoveryConfig
+}
+
+// Create the dkv lookup key from region info
+// Since granularity of an update is at a region per node, the same is used for creating a key
+// As lookups will be mostly on database and possibly on vBucket (within a database) the same order is used for key
+// This enables faster scans for database/vBucket
+func createKeyToInsert(info *RegionInfo) []byte {
+	return []byte(info.GetDatabase() + ":" + info.GetVBucket() + ":" + info.GetNodeAddress())
+}
+
+// Create lookup key based on input request
+// Using the sorted nature of keys, lookup by either database or by database and vBucket or all keys
+func createKeyToGet(request *GetClusterInfoRequest) []byte {
+	if request.GetDatabase() != "" {
+		if request.GetVBucket() != "" {
+			return []byte(request.GetDatabase() + ":" + request.GetVBucket() + ":");
+		} else {
+			return []byte(request.GetDatabase() + ":");
+		}
+	} else {
+		return []byte{};
+	}
+}
+
+func (d *discoverService) UpdateStatus(ctx Context, request *UpdateStatusRequest) (*Status, error) {
+	// Store the serialised form of the request
+	val, err := json.Marshal(request);
+
+	if err != nil {
+		d.logger.Error("Unable to marshal status request", zap.Error(err));
+		return newErrorStatus(err), err;
+	}
+
+	// TODO - check if expiry TTL is working
+	putRequest := PutRequest{Key: createKeyToInsert(request.GetRegionInfo()), Value: val, ExpireTS: hlc.GetUnixTimeFromNow(d.config.statusTTl)};
+
+	if _, err := d.dkvServer.Put(ctx, &putRequest); err != nil {
+		return newErrorStatus(err), err;
+	} else {
+		return newEmptyStatus(), nil;
+	}
+}
+
+func (d *discoverService) GetClusterInfo(ctx Context, request *GetClusterInfoRequest) (*GetClusterInfoResponse, error) {
+	// Create a prefix multi get request to loop through all required status updates
+	// Iterator sends the responses on grpc stream which is unnecessary here
+	getRequest := PrefixMultiGetRequest{KeyPrefix: createKeyToGet(request)}
+
+	if clusterInfo, err := d.dkvServer.PrefixMultiGet(ctx, &getRequest); err == nil {
+		regionsInfo := []*RegionInfo{};
+		i := 0;
+		for _, serializedStatusUpdate := range clusterInfo.KeyValues {
+			statusUpdate := UpdateStatusRequest{};
+			if err := json.Unmarshal(serializedStatusUpdate.Value, &statusUpdate); err == nil {
+				// Filter regions within the requested DC (if provided)
+				if (request.GetDcID() == "" || request.GetDcID() == statusUpdate.GetRegionInfo().GetDcID()) {
+					// Filter inactive regions and regions whose status was updated long time back and hence considered inactive
+					if (hlc.GetTimeAgo(statusUpdate.GetTimestamp()) < d.config.heartbeatTimeout && statusUpdate.GetRegionInfo().GetStatus() != RegionStatus_INACTIVE) {
```

Comment:
Please flatten this nested if. Nested loops are harder to read and maintain. Use `continue` if you want to skip the execution at any point